gapic/src/main/com/google/gapid/util/Logging.java

```diff
@@ -223,52 +255,98 @@ public void close() {
   }
 
   /**
-   * {@link Handler} that keeps a number of messages in memory for later retrieval.
+   * MessageIterator is an iterator for the {@link RingBuffer}.
+   * As the {@link RingBuffer} is a fixed size ring buffer, messages can be discarded if the buffer
+   * fills more quickly than it is consumed. This iterator has been implemented to read the next
+   * most recently available messages. Some messages may be skipped if the iterator lags behind
+   * new messages.
    */
-  private static class MemoryHandler extends Handler {
-    private static final int MAX_SIZE = 10 * 1024 * 1024;
+  public static class MessageIterator {
+    private final RingBuffer target;
+    private int generation;
 
-    public Runnable listener;
-    private final StringBuilder buffer = new StringBuilder();
+    MessageIterator(RingBuffer target) {
+      this.target = target;
+    }
 
-    public MemoryHandler() {
-      listener = () -> { /* empty */ };
+    public Log.Message next() {
+      synchronized (target) {
+        int remaining = Math.min(target.generation - generation, target.ring.length);
+        if (remaining == 0) {
+          return null;
+        }
+        // Progress the generation so that it is at least the oldest message.
+        generation = target.generation - remaining;
+        Log.Message message = target.ring[generation % target.ring.length];
+        generation++;
+        return message;
+      }
     }
+  }
 
-    public synchronized String getMessages() {
-      return buffer.toString();
+  /**
+   * A ring buffer of {@link LogRecord}s.
+   */
+  protected static class RingBuffer {
+    protected final Log.Message[] ring;
+    protected int generation;
+
+    public RingBuffer(int maxSize) {
+      this.ring = new Log.Message[maxSize];
+    }
+
+    public synchronized void add(Log.Message message) {
+      ring[generation % ring.length] = message;
+      generation++;
     }
 
+    public MessageIterator iterator() {
+      return new MessageIterator(this);
+    }
+  }
+
+  /**
+   * {@link Handler} that handles {@link LogRecord}s converting them to
+   * {@link com.google.gapid.proto.log.Log.Message}s and then broadcasting them to listeners.
+   */
+  protected static class Handler extends java.util.logging.Handler {
```

Comment:
They both sound like the Proto is the source when it's the target. Opted for `LogToMessageHandler`.